src/cluster/src/communication.rs

```diff
@@ -139,237 +182,305 @@ where
+///
+/// Epoch values also include a random component, to ensure two values produced by different calls
+/// to `Epoch::mint` never compare as equal.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
+struct Epoch {
+    time: u64,
+    nonce: u64,
+}
 
-        let from_listener = match my_index.cmp(&peer_index) {
-            Ordering::Less => true,
-            Ordering::Equal => panic!("someone claimed to be us"),
-            Ordering::Greater => false,
-        };
+impl Epoch {
+    fn mint() -> Self {
+        let time = SystemTime::UNIX_EPOCH
+            .elapsed()
+            .expect("current time is after 1970")
+            .as_millis()
+            .try_into()
+            .expect("fits");
+        let nonce = rand::random();
+        Self { time, nonce }
+    }
 
-        if from_listener {
-            futs.push({
-                let f = async {
-                    await_connection(&listener, my_index, my_epoch)
-                        .await
-                        .map(|(stream, peer_index)| ConnectionEstablished { peer_index, stream })
-                }
-                .boxed();
-                f
-            });
-        }
+    async fn read(s: &mut Stream) -> std::io::Result<Self> {
+        let time = s.read_u64().await?;
+        let nonce = s.read_u64().await?;
+        Ok(Self { time, nonce })
+    }
 
-        let old = std::mem::replace(&mut results[usize::cast_from(peer_index)], Some(stream));
-        if old.is_some() {
-            panic!("connected to peer {peer_index} multiple times");
-        }
+    async fn write(&self, s: &mut Stream) -> std::io::Result<()> {
+        s.write_u64(self.time).await?;
+        s.write_u64(self.nonce).await?;
+        Ok(())
     }
+}
 
-    Ok(results)
+impl fmt::Display for Epoch {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "({}, {})", self.time, self.nonce)
+    }
 }
 
-#[derive(Debug)]
-/// This task can never successfully boot, since
-/// a peer has seen a higher epoch from `environmentd`.
-pub struct EpochMismatch {
-    /// The epoch we know about
-    mine: ClusterStartupEpoch,
-    /// The higher epoch from our peer
-    peer: ClusterStartupEpoch,
+/// Creates socket connections from a list of host addresses.
+///
+/// The item at index i in the resulting vec, is a connection to process i, except for item
+/// `my_index` which is None (no socket to self).
+async fn create_sockets(
+    my_index: usize,
+    addresses: &[String],
+) -> Result<Vec<Option<Stream>>, CreateSocketsError> {
+    let my_address = &addresses[my_index];
+
+    // Binding to a TCP address of the form `hostname:port` unnecessarily involves a DNS query. We
+    // should get the port from here, but otherwise just bind to `0.0.0.0`.
+    let port_re = Regex::new(r":(\d{1,5})$").unwrap();
+    let listen_address = match port_re.captures(my_address) {
+        Some(cap) => format!("0.0.0.0:{}", &cap[1]),
```

Comment:
Just pointing out that this may hardcode IPv4.

Reply:
That's true, but not sure if it's a problem. Also it's not new in this PR, so I'm not planning to change it here. Can do as a follow up, if we think it's necessary/worthwhile.